Fix handling of short discriminator in QRCodeSetupPayloadGenerator

ksperling-apple opened the pull request:
If a SetupPayload contains a short discriminator then
- isValidQRCodePayload() should return false
- trying to generate a QR Code should return INVALID_ARGUMENT
- generating with SetAllowInvalidPayload(true) should work (not die)

Also make `IsCommonTag`, `IsVendorTag`, and `getOptionalVendorData(tag, &info)` public.

The first two are just encoding spec rules that are useful for clients, and the
latter allows clients to read vendor data by tag instead of having to read the
whole list.

Also use default values instead of explicit constructors for OptionalQRCodeInfo
and fix up some doc comments to correctly reference the vendor tag range.

ksperling-apple committed:
Fix handling of short discriminator in QRCodeSetupPayloadGenerator

If a SetupPayload contains a short discriminator then
- isValidQRCodePayload() should return false
- trying to generate a QR Code should return INVALID_ARGUMENT
- generating with SetAllowInvalidPayload(true) should work (not die)

ksperling-apple committed:
SetupPayload tweaks

Make IsCommonTag, IsVendorTag, and getOptionalVendorData(tag, &info) public.

The first two are just encoding spec rules that are useful for clients, and the
latter allows clients to read vendor data by tag instead of having to read the
whole list.

Also use default values instead of explicit constructors for OptionalQRCodeInfo
and fix up some doc comments to correctly reference the vendor tag range.

andy31415 commented on src/setup_payload/QRCodeSetupPayloadGenerator.cpp:
```diff
@@ -173,8 +173,13 @@ static CHIP_ERROR generateBitSet(PayloadContents & payload, MutableByteSpan & bi
+    // If the payload is valid then discriminator will always be long, but we shouldn't die if we're
+    // encoding a short one due to mAllowInvalidPayload being set. Just use the short value in that case.
+    auto const & pd        = payload.discriminator;
+    uint16_t discriminator = (pd.IsShortDiscriminator()) ? pd.GetShortValue() : pd.GetLongValue();
```
```suggestion
    uint16_t discriminator = pd.IsShortDiscriminator() ? pd.GetShortValue() : pd.GetLongValue();
```

andy31415 commented on src/setup_payload/QRCodeSetupPayloadGenerator.cpp:
```diff
@@ -173,8 +173,13 @@ static CHIP_ERROR generateBitSet(PayloadContents & payload, MutableByteSpan & bi
-                                      kPayloadDiscriminatorFieldLengthInBits, kTotalPayloadDataSizeInBits));
+
+    // If the payload is valid then discriminator will always be long, but we shouldn't die if we're
+    // encoding a short one due to mAllowInvalidPayload being set. Just use the short value in that case.
```
should we in this case validate that mAllowInvalidPayload is set before trying to use the short value?

Please add a description of the use case of when we would like to encode invalid/incomplete data to the comment here. This just says "we shoudn't die" (replace this "shouldn't return error" probably) but it does not explain why. It should explain why.

bzbarsky-apple replied:
"shouldn't die" means "shouldn't die": `GetLongValue()` involves a VerifyOrDie that the value is long.

ksperling-apple replied:
> should we in this case validate that mAllowInvalidPayload is set before trying to use the short value

If `mAllowInvalidPayload` is not set, then we won't get here with a short discriminator because this is now checked by `isValidQRCodePayload`.

My use case here is to allow maybe-invalid payloads to be serialized, for the purpose of deserializing them again. "AllowInvalidPayload" already describes this pretty well?

andy31415 commented on src/setup_payload/SetupPayload.h:
```diff
@@ -153,29 +153,19 @@ enum optionalQRCodeInfoType
-    int64_t int64;
-    uint64_t uint32;
-    uint64_t uint64;
+    int64_t int64   = 0; /**< the integer value if type is optionalQRCodeInfoTypeInt64, otherwise should not be set */
```
not for your PR, however I think we should start using variants here or something like that. I imagine std::string could not be placed in a union with int32 above, so now we have i64/u32/u64 added on top of that again without a union even though only a subset of them can be valid. This feels very brittle.

bzbarsky-apple commented on src/setup_payload/QRCodeSetupPayloadGenerator.cpp:
```diff
@@ -173,8 +173,13 @@ static CHIP_ERROR generateBitSet(PayloadContents & payload, MutableByteSpan & bi
+    // If the payload is valid then discriminator will always be long, but we shouldn't die if we're
+    // encoding a short one due to mAllowInvalidPayload being set. Just use the short value in that case.
+    auto const & pd        = payload.discriminator;
+    uint16_t discriminator = (pd.IsShortDiscriminator()) ? pd.GetShortValue() : pd.GetLongValue();
```
So here's the thing: this payload is not just "invalid", it's literally nonsensical.  Does anything (including tests, etc, given that needs explicit opt-in) actually rely on encoding such payloads with short discriminators as QR codes?  If not, I don't think we should support it and should just error out.  The resulting QR code is not "invalid" in any sense, so it can't be used to test handling of invalid QR codes by other entities; it's just not a valid QR code that matches any reality in any way.

ksperling-apple replied:
My specific use case is to use the QR code representation to serialize MTRSetupPayload objects that may not be valid, and be able to deserialize them again. Using the QR code TLV representation seems like a much more robust way to handle the extension data than coming up with some other encoding for it.

ksperling-apple replied:
As discussed in Slack, I'm going to just encode the long discriminator as 0, to avoid a client making any undue assumptions about being able to roundtrip a short discriminator here.

ksperling-apple committed:
Address review comments

Elaborate on the use case for AllowInvalidPayload in the comments, and encode a
missing long discriminator as 0, to avoid a client encoding invalid payloads
from relying on round-tripping a short discriminator through a QR code.

ksperling-apple committed:
Handle rendezvousInformation and discriminator the same way

andy31415 commented on src/setup_payload/QRCodeSetupPayloadGenerator.cpp:
```diff
@@ -170,10 +175,11 @@ static CHIP_ERROR generateBitSet(PayloadContents & payload, MutableByteSpan & bi
-                                      kRendezvousInfoFieldLengthInBits, kTotalPayloadDataSizeInBits));
-    ReturnErrorOnFailure(populateBits(bits.data(), offset, payload.discriminator.GetLongValue(),
+    auto const & ri = payload.rendezvousInformation;
+    ReturnErrorOnFailure(populateBits(bits.data(), offset, (ri.HasValue() ? ri.Value().Raw() : 0), kRendezvousInfoFieldLengthInBits,
```
I wonder if this is more readable:

```cpp
    ReturnErrorOnFailure(populateBits(bits.data(), offset, ri.ValueOr(RendezvousInformationFlag::kNone).Raw(),
                                      kRendezvousInfoFieldLengthInBits, kTotalPayloadDataSizeInBits));
```

or 

```cpp
    ReturnErrorOnFailure(populateBits(bits.data(), offset,
                                      payload.rendezvousInformation.ValueOr(RendezvousInformationFlag::kNone).Raw(),
                                      kRendezvousInfoFieldLengthInBits, kTotalPayloadDataSizeInBits));
```

ksperling-apple replied:
Done

ksperling-apple committed:
Address review comment: use ValueOr
